codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/SerializeImplGenerator.kt

```diff
@@ -0,0 +1,392 @@
+import software.amazon.smithy.rust.codegen.core.smithy.isOptional
+import software.amazon.smithy.rust.codegen.core.smithy.protocols.shapeFunctionName
+import software.amazon.smithy.rust.codegen.core.smithy.rustType
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.hasTrait
+import software.amazon.smithy.rust.codegen.core.util.isTargetUnit
+import software.amazon.smithy.rust.codegen.core.util.letIf
+import software.amazon.smithy.rust.codegen.core.util.toPascalCase
+
+class SerializeImplGenerator(private val codegenContext: CodegenContext) {
+    fun generateRootSerializerForShape(shape: Shape): Writable = serializerFn(shape, null)
+
+    /**
+     * Generates a serializer for a given shape. Collection serializer require using a wrapper structure. To handle this,
+     * `applyTo` allows passing in a `Writable` referring to an input. This returns a writable that wraps the input if necessary,
+     * e.g.
+     *
+     * `&self.map_shape` => `SerializeSomeSpecificMap(&self.map_shape)`
+     *
+     * Even if no wrapping is required, the writable that is returned includes the required serializer as dependencies.
+     */
+    private fun serializerFn(
+        shape: Shape,
+        applyTo: Writable?,
+    ): Writable {
+        val name = codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape) + "_serde"
+        val deps =
+            when (shape) {
+                is StructureShape -> RuntimeType.forInlineFun(name, SerdeModule, structSerdeImpl(shape))
+                is UnionShape -> RuntimeType.forInlineFun(name, SerdeModule, serializeUnionImpl(shape))
+                is TimestampShape -> serializeDateTime(shape)
+                is BlobShape -> serializeBlob(shape)
+                is StringShape, is NumberShape -> directSerde(shape)
+                is DocumentShape -> serializeDocument(shape)
+                else -> null
+            }
+        return writable {
+            val wrapper =
+                when {
+                    deps != null -> null
+                    shape is MapShape -> serializeMap(shape)
+                    shape is CollectionShape -> serializeList(shape)
+                    // Need to figure out the best default here.
+                    else -> PANIC("No serializer supported for $shape")
+                }
+            if (wrapper != null && applyTo != null) {
+                rustTemplate("&#{wrapper}(#{applyTo})", "wrapper" to wrapper, "applyTo" to applyTo)
+            } else {
+                deps?.toSymbol().also { addDependency(it) }
+                applyTo?.invoke(this)
+            }
+        }
+    }
+
+    private fun serializeMap(shape: MapShape): RuntimeType =
+        serializeWithWrapper(shape) { value ->
+            writable {
+                rustTemplate(
+                    """
+                    use #{serde}::ser::SerializeMap;
```

Comment:
nit: could use FQS instead of having these `use` statements: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name